### DISCUSSION TITLE
How can I update the URL in browser address bar when using middleware redirects in Next 13?

### DISCUSSION THREAD

**Discussion by mryechkin** [vercel/next.js#46445] [Help] [answered]
In my middleware I'm checking if the given route has a session present, and if not - redirect to the `/sign-in` page:

```
// src/middleware.js
import { NextResponse } from 'next/server';

export async function middleware(req) {
  const res = NextResponse.next();

  // check for session here

  if (!session) {
    return NextResponse.redirect(new URL('/sign-in', req.url));
  }

  return res;
}

export const config = {
  matcher: ['/', '/profile'],
};
```

Now, it works as expected - but the URL in the browser stays the same as the original route. So for example if I'm not logged-in and go to `/profile`, I'm redirected to my `/sign-in` page, but the URL stays as `/profile`. Same thing with the root route.

Is there a way to make the URL change as well after the redirect? I was always under the impression that was supposed to be the default behaviour. Am I just missing something obvious? Would appreciate any help. Thanks

**Comment by icyJoseph** [comment, marked as answer]
Please share a repository if possible. What you are describing is the behavior of rewrite, rather than redirect.

Are you sure you don't have rewrite instead? Or perhaps have next.config.js rewrites setup? Kind of confusing situation you are in.

**Reply by mryechkin** [reply]
Thanks for confirming. In trying to reproduce the issue in a blank repo I can see that it does indeed work as expected. I'm also using [router.refresh()](https://beta.nextjs.org/docs/api-reference/use-router#userouter) in my project, so must be something there that's interfering. I'll keep investigating

**Reply by icyJoseph** [reply]
If you manage to isolate it, it would be nice to take a look at an example repository. Maybe there is still something buggy with Middleware and hard/soft app dir navigation.

**Comment by jvnkr** [comment]
If you still have this problem you can try hard refreshing using `window.location.href` instead of `router.refresh()`.
